File: Backend/services/aiService.js
/**
<<<<<<< Updated upstream
 * AI Service — stub implementations.
 * Replace with real LLM calls (OpenAI / Azure OpenAI) when ready.
 */

/**
 * General-purpose AI chat.
 */
const chat = async ({ message, context }) => {
  // TODO: integrate with OpenAI / Azure OpenAI
  return {
    role: 'assistant',
    text: `Based on the available clinical data, here is my analysis:\n\n` +
          `You asked: "${message}"\n\n` +
          `This is a simulated AI response. In production, this will be powered by a clinical LLM ` +
          `with access to the patient's EMR data for evidence-based suggestions.\n\n` +
          `Key capabilities:\n` +
          `• Clinical decision support\n` +
          `• Drug interaction checking\n` +
          `• Differential diagnosis suggestions\n` +
          `• Patient history summarization`,
    time: new Date().toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
  };
};

/**
 * Generate SOAP note from transcription.
 */
const generateSoapNote = async ({ transcription, patient_context }) => {
  // TODO: integrate with clinical LLM
  return {
    subjective: 'Patient reports chest discomfort during physical activity, specifically when climbing stairs. ' +
                'Describes the sensation as pressure-like in the center of the chest, resolving with rest within 2-3 minutes. ' +
                'Also notes occasional tingling in the left arm. Has been compliant with medications (Metformin, Atorvastatin, Nitroglycerin). ' +
                'Used Nitroglycerin twice this week for symptom relief.',
    objective:  'BP: 138/88 mmHg (slightly elevated). HR: 82 bpm. ' +
                'Patient appears in no acute distress. Cardiovascular exam pending stress test results.',
    assessment: 'Stable angina pectoris with ongoing exertional symptoms. ' +
                'Hypertension - borderline, needs monitoring. ' +
                'Type 2 Diabetes - on current regimen.',
    plan:       '1. Order cardiac stress test to evaluate heart function under exertion.\n' +
                '2. Review most recent lab results (lipid panel, HbA1c).\n' +
                '3. Continue current medications.\n' +
                '4. Follow-up in 2 weeks for stress test results.\n' +
                '5. Patient education: report any worsening symptoms or chest pain at rest immediately.',
  };
};

/**
 * Generate ICD-10 codes from diagnosis/transcription.
 */
const generateIcdCodes = async ({ transcription, diagnosis }) => {
  // TODO: integrate with clinical LLM / ICD code database
  return [
    { code: 'I20.9',  description: 'Angina pectoris, unspecified' },
    { code: 'I10',    description: 'Essential (primary) hypertension' },
    { code: 'E11.9',  description: 'Type 2 diabetes mellitus without complications' },
    { code: 'R07.9',  description: 'Chest pain, unspecified' },
  ];
};

module.exports = { chat, generateSoapNote, generateIcdCodes };
=======
 * AI Service — Proxy to the FastAPI AI Backend
 * 
 * Forwards requests to the Python FastAPI server for:
 *   - Audio transcription (Whisper + speaker diarization)
 *   - Medical entity extraction (MedGemma)
 *   - EMR record generation (MedGemma)
 *   - Patient summary generation (MedGemma)
 *   - ICD code lookup
 *   - Differential diagnosis suggestions
 */

const axios = require('axios');
const FormData = require('form-data');
const config = require('../config/config');

const AI_BASE_URL = config.ai.baseUrl;

/**
 * POST /api/v1/transcribe — Upload audio → transcription + speaker classification
 * @param {Buffer} fileBuffer - Audio file buffer
 * @param {string} filename  - Original filename
 * @param {string} mimetype  - MIME type (audio/wav, audio/mp3, etc.)
 * @returns {{ transcript, audio_duration_seconds }}
 */
async function transcribeAudio(fileBuffer, filename, mimetype) {
  const form = new FormData();
  form.append('file', fileBuffer, { filename, contentType: mimetype });

  const { data } = await axios.post(`${AI_BASE_URL}/transcribe`, form, {
    headers: form.getHeaders(),
    maxContentLength: Infinity,
    maxBodyLength: Infinity,
    timeout: 300_000, // 5 min — transcription can be slow
  });
  return data;
}

/**
 * POST /api/v1/extract — Transcript → medical entities (vitals, symptoms, dx, meds…)
 * @param {object} transcript - ClassifiedTranscript object
 * @returns {{ entities }}
 */
async function extractEntities(transcript) {
  const { data } = await axios.post(`${AI_BASE_URL}/extract`, { transcript }, {
    timeout: 120_000,
  });
  return data;
}

/**
 * POST /api/v1/generate-emr — Entities + transcript → structured EMR record
 * @param {object} params
 * @param {object} params.entities   - ExtractedEntities
 * @param {object} params.transcript - ClassifiedTranscript
 * @param {string} [params.patient_id]
 * @param {string} [params.encounter_date]
 * @param {string} [params.encounter_type]
 * @param {number} [params.audio_duration]
 * @param {string} [params.patient_name]
 * @param {string} [params.provider_name]
 * @param {string} [params.facility_name]
 * @returns {{ emr_record }}
 */
async function generateEMR(params) {
  const { data } = await axios.post(`${AI_BASE_URL}/generate-emr`, params, {
    timeout: 120_000,
  });
  return data;
}

/**
 * POST /api/v1/patient-summary — EMR record → patient-friendly summary
 * @param {object} emrRecord - Full EMR record object
 * @returns {{ patient_summary: string }}
 */
async function generatePatientSummary(emrRecord) {
  const { data } = await axios.post(`${AI_BASE_URL}/patient-summary`, {
    emr_record: emrRecord,
  }, {
    timeout: 120_000,
  });
  return data;
}

/**
 * POST /api/v1/icd-lookup — Fuzzy search for ICD codes
 * @param {string} query   - Symptom or condition text
 * @param {number} [version=10] - ICD version (9 or 10)
 * @param {number} [topK=5]
 * @returns {{ query, matches }}
 */
async function icdLookup(query, version = 10, topK = 5) {
  const { data } = await axios.post(`${AI_BASE_URL}/icd-lookup`, {
    query,
    version,
    top_k: topK,
  }, {
    timeout: 30_000,
  });
  return data;
}

/**
 * POST /api/v1/suggest-diagnoses — AI differential diagnosis with ICD mapping
 * @param {object} params
 * @param {Array}  params.symptoms
 * @param {string} [params.age]
 * @param {string} [params.gender]
 * @param {object} [params.social_history]
 * @param {Array}  [params.family_history]
 * @returns {{ suggestions }}
 */
async function suggestDiagnoses(params) {
  const { data } = await axios.post(`${AI_BASE_URL}/suggest-diagnoses`, params, {
    timeout: 120_000,
  });
  return data;
}

/**
 * GET /api/v1/health — Check AI backend health
 */
async function checkHealth() {
  const { data } = await axios.get(`${AI_BASE_URL}/health`, { timeout: 10_000 });
  return data;
}

module.exports = {
  transcribeAudio,
  extractEntities,
  generateEMR,
  generatePatientSummary,
  icdLookup,
  suggestDiagnoses,
  checkHealth,
};
>>>>>>> Stashed changes

File: Backend/app/services/ai_service.py
"""
AI Service — HTTP client for the AI Backend (merged FastAPI server)
"""

import logging
from typing import Any, Dict, Optional

import httpx

from app.config import settings

logger = logging.getLogger(__name__)

_client: Optional[httpx.AsyncClient] = None


def _get_client() -> httpx.AsyncClient:
    global _client
    if _client is None:
        _client = httpx.AsyncClient(base_url=settings.AI_BACKEND_URL, timeout=600.0)
    return _client


async def health_check() -> Dict[str, Any]:
    """Check AI backend health."""
    try:
        resp = await _get_client().get("/health")
        return resp.json()
    except Exception as e:
        logger.warning(f"AI backend health check failed: {e}")
        return {"status": "unreachable", "error": str(e)}


async def _safe_post(path: str, **kwargs) -> Dict[str, Any]:
    """POST to AI backend with graceful error handling."""
    client = _get_client()
    try:
        resp = await client.post(path, **kwargs)
        resp.raise_for_status()
        return resp.json()
    except httpx.ConnectError:
        logger.warning(f"AI backend unreachable for {path}")
        return {"error": "AI backend is not running", "status": "unavailable"}
    except Exception as e:
        logger.error(f"AI backend error for {path}: {e}")
        return {"error": str(e), "status": "failed"}


async def transcribe_audio(audio_bytes: bytes, filename: str) -> Dict[str, Any]:
    """Send audio to AI backend for transcription."""
    return await _safe_post("/transcribe", files={"file": (filename, audio_bytes, "audio/mpeg")})


async def extract_entities(transcript: str) -> Dict[str, Any]:
    """Extract medical entities from transcript."""
    return await _safe_post("/extract", json={"transcript": transcript})


async def generate_emr(extraction_data: Dict[str, Any]) -> Dict[str, Any]:
    """Generate full EMR from extracted entities."""
    return await _safe_post("/generate-emr", json={"extraction": extraction_data})


async def generate_patient_summary(emr_data: Dict[str, Any]) -> Dict[str, Any]:
    """Generate patient-friendly summary."""
    return await _safe_post("/patient-summary", json={"emr_record": emr_data})


async def icd_lookup(query: str, version: int = 10, limit: int = 10) -> Dict[str, Any]:
    """Search ICD codes."""
    return await _safe_post("/icd-lookup", json={"query": query, "version": version, "limit": limit})


async def suggest_diagnoses(symptoms: str) -> Dict[str, Any]:
    """Get AI-suggested diagnoses."""
    # The AI backend expects symptoms as a list of Symptom objects
    symptoms_list = [{"description": s.strip()} for s in symptoms.split(",")] if symptoms else []
    return await _safe_post("/suggest-diagnoses", json={"symptoms": symptoms_list})


async def analyze_document_with_ai(
    file_bytes: bytes, filename: str, content_type: str = "application/pdf",
) -> Dict[str, Any]:
    """Send document to AI backend for analysis (doc analysis pipeline)."""
    return await _safe_post("/doc/analyze-document", files={"file": (filename, file_bytes, content_type)})


async def analyze_document_for_patient(
    file_bytes: bytes, filename: str, content_type: str = "application/pdf",
) -> Dict[str, Any]:
    """Patient-facing document analysis: structured data + patient-friendly summary.
    Faster than full analysis (1 Qwen + 1 MedGemma call).
    """
    return await _safe_post(
        "/doc/patient-report",
        files={"file": (filename, file_bytes, content_type)},
    )


async def analyze_document_for_clinician(
    file_bytes: bytes,
    filename: str,
    content_type: str = "application/pdf",
    medical_history: Optional[str] = None,
) -> Dict[str, Any]:
    """Clinician-facing document analysis: clinical insights, ICD codes, flags, recommendations.
    Accepts optional medical_history JSON for deeper contextual analysis.
    """
    data = {"medical_history": medical_history} if medical_history else {}
    return await _safe_post(
        "/doc/clinician-report",
        files={"file": (filename, file_bytes, content_type)},
        data=data,
    )


async def chat_with_medgemma(messages: list) -> str:
    """Chat with MedGemma via the AI backend."""
    client = _get_client()
    try:
        resp = await client.post("/doc/chat", json={"messages": messages})
        resp.raise_for_status()
        data = resp.json()
        return data.get("response", data.get("content", ""))
    except Exception as e:
        logger.error(f"MedGemma chat failed: {e}")
        return f"AI service unavailable: {e}"
